Sources/SmokeHTTP1/SmokeHTTP1Server.swift

```diff
@@ -89,20 +260,83 @@ public class SmokeHTTP1Server {
             .childChannelOption(ChannelOptions.allowRemoteHalfClosure, value: true)
         
         channel = try bootstrap.bind(host: ServerDefaults.defaultHost, port: port).wait()
+        
+        fullyShutdownPromise.futureResult.whenComplete { [unowned self] in
+            do {
+                let shutdownCompletionHandlers = self.lifecycleState.updateStateOnShutdownComplete()
+                
+                // execute all the completion handlers
+                shutdownCompletionHandlers.forEach { self.shutdownCompletionHandlerInvocationStrategy.invoke(handler: $0) }
+                
+                try self.group.syncShutdownGracefully()
+                
+                // release any waiters for shutdown
+                self.shutdownSemaphore.signal()
+            } catch {
+                Log.error("Server unable to shutdown cleanly following full shutdown.")
+            }
+            
+            Log.info("SmokeHTTP1Server shutdown.")
+        }
+        
+        Log.info("SmokeHTTP1Server started on port \(port).")
     }
     
     /**
-     Stops the server.
+     Initiates the process of shutting down the server.
      */
-    public func stop() throws {
-        try group.syncShutdownGracefully()
-        try threadPool.syncShutdownGracefully()
+    public func shutdown() throws {
+        let doShutdownServer = try lifecycleState.updateOnShutdownStart()
+        
+        if doShutdownServer {
+            quiesce.initiateShutdown(promise: fullyShutdownPromise)
+        }
+    }
+    
+    /**
+     Blocks until the server has been shutdown and all completion handlers
+     have been executed.
+     */
+    public func waitUntilShutdown() throws {
+        if !lifecycleState.isShutdown() {
+            shutdownSemaphore.wait()
+        }
     }
     
     /**
-     Blocks until the server has been shut down.
+     Blocks until the server has been shutdown and all completion handlers
+     have been executed. The provided closure will be added to the list of
+     completion handlers to be executed on shutdown. If the server is already
+     shutdown, the provided closure will be immediately executed.
+     
+     - Parameters:
+        - onShutdown: the closure to be executed after the server has been
+                      fully shutdown.
      */
-    public func wait() throws {
-        try channel?.closeFuture.wait()
+    public func waitUntilShutdownAndThen(onShutdown: @escaping () -> Void) throws {
+        let handlerQueuedForFutureShutdownComplete = lifecycleState.addShutdownHandler(onShutdown: onShutdown)
```

Comment:
these `wait`s aren't quite correct as there'll only ever be a single `signal`. What you can do is:

    let sem = DispatchSemaphore(value: 0) // this semaphore will always be signalled once and waited on once
    let handlerQueuedForFutureShutdownComplete = lifecycleState.addShutdownHandler {
        onShutdown()
        sem.signal()
    }
    
    if !handlerQueuedForFutureShutdownComplete {
        shutdownCompletionHandlerInvocationStrategy.invoke(handler: onShutdown)
        sem.signal()
    }
    sem.wait()

Reply:
and `DispatchGroup` is probably a much nicer interface for that. You can do `group.notify` to be notified and `group.wait` to wait on it and it should do everything else for you. Ie. you can wait multiple times

```swift
import Dispatch

let q = DispatchQueue(label: "Q")
let g = DispatchGroup()

g.enter() // when committed to do the shutdown
g.leave() // when the shutdown completed

g.wait() // you
g.wait() // can
g.wait() // wait
g.wait() // as
g.wait() // many
g.wait() // times
g.wait() // as you want
g.notify(queue: q) { // or get notified
    print("hi")
}
g.notify(queue: q) {
    print("hi")
}

g.notify(queue: q) {
    print("hi")
}
```

Reply:
Thanks for all this advice!

I have looked into using DispatchGroup based on this and am considering keeping `shutdownCompletionHandlers` for the list of shutdown completion handlers and changing the default `shutdownCompletionHandlerInvocationStrategy` to invoke the completion handlers synchronously. This would mean that `waitUntilShutdown*` will not unblock until all completion handlers have returned. I couldn't find any documentation on how `DispatchGroup.notify` are invoked in relation to waiters on the group.

Is it advisable that `fullyShutdownPromise.futureResult.whenComplete` will potentially block on the shutdown completion handlers before releasing any waiters on the group to provide this guaranteed order of execution as an option? Or is it safer not to provide any guaranteed order of execution between the completion handlers and unblocking any waiters?